crates/integrations/datafusion/src/table/table_provider_factory.rs

```diff
@@ -0,0 +1,312 @@
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+#[async_trait]
+impl TableProviderFactory for IcebergTableProviderFactory {
+    async fn create(
+        &self,
+        _state: &dyn Session,
+        cmd: &CreateExternalTable,
+    ) -> DFResult<Arc<dyn TableProvider>> {
+        check_cmd(cmd).map_err(to_datafusion_error)?;
+
+        let table_name = &cmd.name;
+        let metadata_file_path = &cmd.location;
+        let options = &cmd.options;
+
+        let table_name_with_ns = complement_namespace_if_necessary(table_name);
+
+        let table = create_static_table(table_name_with_ns, metadata_file_path, options)
+            .await
+            .map_err(to_datafusion_error)?
+            .into_table();
+
+        let schema = schema_to_arrow_schema(table.metadata().current_schema())
+            .map_err(to_datafusion_error)?;
+
+        Ok(Arc::new(IcebergTableProvider::new(table, Arc::new(schema))))
+    }
+}
+
+fn check_cmd(cmd: &CreateExternalTable) -> Result<()> {
+    println!("Checking command: {:?}", cmd);
+
+    let CreateExternalTable {
+        schema,
+        table_partition_cols,
+        order_exprs,
+        constraints,
+        column_defaults,
+        ..
+    } = cmd;
+
+    // Check if any of the fields violate the constraints in a single condition
+    let is_invalid = !schema.fields().is_empty()
+        || !table_partition_cols.is_empty()
+        || !order_exprs.is_empty()
+        || !constraints.is_empty()
+        || !column_defaults.is_empty();
+
+    if is_invalid {
+        return Err(Error::new(ErrorKind::FeatureUnsupported, "Currently we only support reading existing icebergs tables in external table command. To create new table, please use catalog provider."));
+    }
+
+    Ok(())
+}
+
+/// Complements the namespace of a table name if necessary.
+///
+/// # Example
+///
+/// ```
+/// use datafusion::sql::TableReference;
+/// use iceberg_datafusion::complement_namespace_if_necessary;
+///
+/// let table_name = TableReference::bare("table_name");
+/// let table_name_with_ns = complement_namespace_if_necessary(&table_name);
+///
+/// assert_eq!(
+///     table_name_with_ns,
+///     TableReference::partial("default", "table_name")
+/// );
+/// ```
+///
+/// # Note
+/// If the table name is a bare name, it will be complemented with the 'default' namespace.
+/// Otherwise, it will be returned as is. Because Iceberg tables are always namespaced, but DataFusion
+/// external table commands not include the namespace, this function ensures that the namespace is always present.
+///
+/// # See also
+/// - [`iceberg::NamespaceIdent`]
+/// - [`datafusion::sql::planner::SqlToRel::external_table_to_plan`]
+fn complement_namespace_if_necessary(table_name: &TableReference) -> Cow<'_, TableReference> {
+    match table_name {
+        TableReference::Bare { table } => {
+            Cow::Owned(TableReference::partial("default", table.as_ref()))
+        }
+        other => Cow::Borrowed(other),
+    }
+}
```

Comment:
Please pay attention to this function, because when `DataFusion` creates external tables, it only creates bare `TableReference`, which does not include a `schema`. However, when constructing `TableIdent` in Iceberg, the 'schema' is required. Does anyone have a good approach for handling compatibility?

Reply:
I ran into an issue with external tables only creating bare `TableReference` as well - do you know if thats the intended behavior or if its a bug in datafusion?

Reply:
@matthewmturner I believe this is a known issue, see https://github.com/apache/datafusion/blob/main/datafusion/sql/src/statement.rs#L1242

Reply:
Yeah weve got a [ticket](https://github.com/apache/datafusion/issues/12607) created for that now. thanks @yukkit

Reply:
@liurenjie1024 You’re right. When creating an external table in DataFusion without specifying a schema, it will to being created under the default schema. In this case, the table name passed to `TableProviderFactory` does not include schema information. However, `TableIdent` requires a namespace, so I use "default" to populate this value. Anyway, the `IcebergTableProvider` doesn't actually rely on the specific content of `TableIdent`, so it has no impact when used as an external table in DataFusion (at least as things stand).